### ISSUE TITLE
hyperperameters setting of MAML and test code

### ISSUE THREAD

**Issue by Hugo101** [learnables/learn2learn#135]
For the [MAML reproducing](https://github.com/learnables/learn2learn/tree/master/examples/vision), the document says that **Note that the original MAML paper trains with 5 fast adaptation step, but tests with 10 steps. This implementation only provides the training code.**
So does that mean the test codes in the file [maml_miniimagenet.py](https://github.com/learnables/learn2learn/blob/master/examples/vision/maml_miniimagenet.py) and [maml_omniglot.py](https://github.com/learnables/learn2learn/blob/master/examples/vision/maml_omniglot.py) are not correct? I found that it uses the same function (`fast_adapt`) with training code.

Also, could you please provide the specific hyperperameters setting for the **reproducing table of MAML** for MiniImageNet and Omniglot? I use the default setting and did not change the test code, the results are not good. It says **Only the fast-adaptation learning rate needs a bit of tuning, and good values usually lie in a 0.5-2x range of the original value.** So we only need to fine tune the`fast_lr` in the range [0.5, 2] and use the test code in the current [maml_miniimagenet.py](https://github.com/learnables/learn2learn/blob/master/examples/vision/maml_miniimagenet.py) and [maml_omniglot.py](https://github.com/learnables/learn2learn/blob/master/examples/vision/maml_omniglot.py)? The original value means the `fast_lr=0.5` in the code or the original value in the original fast adaptation lr in the original MAML paper?

I am sorry for these naive questions, I have tried many times and did not get the results in the table, which has token too much much time.

Thank you so much!

**Comment by Kostis-S-Z** [comment]
Hello Hugo!
The code for training and testing i think is correct. The only thing that needs modification is the hyper-parameter values. Personally, I didn't have trouble reproducing the MiniImagetNet results with a bit of fine-tuning. For example, for 5-way 1-shot classification I used exactly the same hyper-parameter values as in maml_miniimagenet.py and I got around 48-50% test accuracy across 3 different seeds. Could you maybe post your results here? How much worse are they compared to the ones in the table?

_As I see it, these examples act more like reference points on how to use the library and might not be perfectly fine-tuned to provide the best results out of the box._

**Comment by Hugo101** [comment]
Hi Kostis,
Thanks for your quick reply!
I have some doubts about testing code. 

1. The number of tasks it tests is only `meta_batch_size (32)`, not the whole meta-testing tasks (1024).

2. It uses  the same `fast_adapt` function with meta-training. Sometimes, yes, it makes sense. Sometimes, it's not enough. For example, 

- during testing, as said by the document, 10 update steps could be used compared with 5 in training. 

- Also, step size could be different from meta-training. 
So using the same `fast_adapt` function is not perfect.

3. For reproducing, 5-way 1-shot of maml_miniimagenet.py seems good without any changes. However, my 5-way 5-shot of that is 56%, which is less than 63%. I need to try different seeds perhaps.

4. BTW, How long will it take for you to run one experiment? It will take over 24 hours using 2080Ti.

Thank you so much!
Hugo

**Comment by Kostis-S-Z** [comment]
> The code for training and testing i think is correct. The only thing that needs modification is the hyper-parameter values.

What I meant by this is that if you modify the values of the hyper-parameters in different places in the code, you can make it work as you want. 

1. For evaluation, instead of using the `meta_batch_size` as the number of tasks you want to evaluate you can just change it for the value that you want e.g 1024. 

2. Same for the function `fast_adapt`. It works just fine for meta-testing as well but you need to make some changes to the arguments. For example, the argument `adaptation_steps`  can be changed from 5 to 10 when calling the function during evaluating. Also the step size (or learning rate) can be easily changed during meta-testing by writing `maml.lr = 0.5 ` after training and before evaluating.

3. For 5-way 5-shot I actually managed to get 61.6%, 63.7% and 64.8% across 3 different seeds (actual seed values were 1, 2 and 3) with the exact same hyper-parameters! I run the experiments for 10.000 or 20.000 iterations. I am not sure why you are getting worse results...

4. I run all the experiments on a GTX 1080 Ti, however I run multiple experiments in parallel in the same GPU so there is some overhead. If I were to run one experiment at a time I would expect slightly better times than the ones below. This is also why the times might not be very consistent.
```
Mini ImageNet 5ways 5shots | 20.000iter | 26hours
Mini ImageNet 5ways 1shot | 20.000iter | 11hours
Omniglot 20ways 5shots | 10.000iter | 29hours
Omniglot 20ways 1shot | 10.000iter | 15hours
```

By the way I am not a developer of this project, just a big fan and user :)

**Comment by Kostis-S-Z** [comment]
No worries, no inconvenience :)

1. Yes, for the results of the experiments I mentioned before thats the only things I changed! 
2. I am not sure if I am misunderstanding your question, but the number of shots in the evaluation set in `fast_adapt` is defined again by the variable `shots` in the `main()` function. For example, you define 5 ways &15 shots (meaning 15 samples / images per class), then during fast_adapt the evaluation set will again contain 5 classes with 15 samples each (in total 5*15=75 samples). This is regardless which dataset you use.

I highly recommend you use a debugger and go step-by-step through the code to get a better understanding!

**Comment by Hugo101** [comment]
Kostis, thank you so much! I am debugging step-by-step.
Sorry for the confusion about the second question. Based on the MAML setting in the original paper, the number of shots in adaptation set and evaluation set (aka **query set** in other meta learning papers) are different usually. 
Take the example of 5-way 5-shot:
5 samples per class is for adaptation set. 15 samples per class is used in evaluation set. As you can see from the paper in **A.1 classification**: **15 examples per class were used for evaluating the post-update meta-gradient.**
In the code, `fast_adapt` only guarantee the shots of adaptation set, the rest is evaluation set no matter the number is.
I am new to this code, I am sorry if my understanding is not correct.
Thanks!

Best,
Hugo

**Comment by Kostis-S-Z** [comment]
I see, yes you are right! This code makes the assumption that the query set will always be composed by N * K samples and all N classes will be represented by K samples. But, as you say, this is not always the case and the query set should be able to be of different size with samples randomly picked from N classes (even if not every class is represented equally).

A quick fix I can think of right now for this would be to change the TaskDataset class to also support a function (let's call it `sample_query(N, L)`) where "_**L**_" (which is different than K) is number of samples for the query set and it would fetch L samples chosen randomly from a pool of N classes. Then you would use the default sample() function (that is currently implemented) to build the _D_train_ set and `sample_query` to build the query set. This would also solve #125 since you would call Kshots with the actual number of shots (not 2*shots). I am not sure if I am missing something right now but this makes sense to me. 

@seba-1511 If you have time take a look at this and let us know what you think. It shouldn't be hard to implement this (would make for a good first issue) and someone could make a PR for this.

(By the way I found the description of the few-shot learning setting more clear in [Optimization as a Model for Few-Shot Learning](https://openreview.net/pdf?id=rJY0-Kcll)", Section 5, than in MAML where they describe this as _"samples for evaluating the post-update meta-gradient"_ which, personally, doesn't sound very intuitive)

**Comment by seba-1511** [comment]
Thanks for starting an interesting conversation @Hugo101 and for great answers @Kostis-S-Z.

I think there are two scenarios we're discussing.

1. Use 5 samples per class for adaptation but 15 samples for evaluation: This requires changing how we assign samples in [fast_adapt](https://github.com/learnables/learn2learn/blob/master/examples/vision/maml_miniimagenet.py#L23). (i.e. manipulating some indices) I think the [ProtoNet example](https://github.com/learnables/learn2learn/blob/master/examples/vision/protonet_miniimagenet.py#L60) already does something similar.
2. Use 5 samples for some classes and 15 samples for others: This requires to define which class gets how many samples. Slightly trickier to implement, but you can roll your own [task transform](http://learn2learn.net/docs/learn2learn.data/#learn2learndatatransforms) to achieve this. Since classes don't have the same number of samples, you'd also need to be careful when assigning data to adaptation and evaluation splits.

Is there an additional case I am missing ? If you need help with the task transform, I'd be happy to provide some information. (Also, making custom task transforms would be a very useful tutorial.)

**Comment by Kostis-S-Z** [comment]
Aha, so I guess the answer can be found in the ProtoNet example!

From my understanding, we were mostly concerned about the first scenario and how we could change the query set size in MAML. And the example you posted solves exactly this problem. So, the changes would be to have another variable for the query set size (`query_num`) and then:
1. Instead of calling KShots like this 
`KShots(train_dataset, 2*shots)`, 
call it like this
`KShots(train_dataset, query_num + args.shot)`
2. Add `query_num` as argument to `fast_adapt`  and inside the function change the code from splitting equally the support and query set to these lines below.
```
    support_indices = np.zeros(data.size(0), dtype=bool)
    selection = np.arange(ways) * (shot + query_num)
    for offset in range(shot):
        support_indices[selection + offset] = True
    query_indices = torch.from_numpy(~support_indices)
    support_indices = torch.from_numpy(support_indices)
```

@Hugo101 Was it this what you were describing?

**Comment by Hugo101** [comment]
Hi Kostis @Kostis-S-Z  and @seba-1511 , sorry for the late reply.
Yes. The first scenario given by @seba-1511  is what I mean, which is also many classic setting in some meta learning/few shot learning papers. As you guys said,  **ProtoNet example** has given some idea about this. @Kostis-S-Z  you also give detailed implementation.

For the second scenario given by Seba, I did not mean that. However, there are some research regarding to this, like **class imbalance / task imbalance** case (or **Any shot classification**) in [this paper](https://openreview.net/pdf?id=rkeZIJBYvr). For specific number of samples for specific class, it's difficult because there are many classes usually, also it's not necessary. 

One question, if considering the second scenario, only dataset (input of the model) changes, is that necessary to modify the MAML model? (This is not related to learn2learn framework. Just general discussion.) Perhaps not necessary.

**Comment by seba-1511** [comment]
No I don't think it'll be necessary to change the MAML code, just how you assign samples to the adaptation and evaluation splits.

I'm closing this issue since it seems we've found a solution.

**Comment by wetliu** [comment]
@Kostis-S-Z Thank you so much for your help! I can't reproduce the mini-imagenet 5 ways 5 shots results to around 63%. 

1) I am wondering for your 5 ways 5 shot training, are these your hyperparameters? 
```
ways=5,
shots=5,
meta_lr=0.003,
fast_lr=0.5,
meta_batch_size=32,
adaptation_steps=1,
num_iterations=20000, # this is changed to 20000
cuda=True,
seed=1 # or 2 or 3
```

2) I notice that adaptation_steps=1, which is quite different from the paper. If my understanding is correct, you did not change this value, leaving adaptation_steps=1. Is this correct? 

3) When you do evaluation (I am using 1024 samples from testset), what is adaptation_steps? Is it 1 (the same as training), or is it 10 (paper reported)?

4) Did you save the "best" model weights through the validation set for evaluation on test set or you just didn't do any evaluations until 20000 steps finish?  

Thank you for your clarification.
